oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,569 @@
+    /// current and desired marker positions.
+    ///
+    /// Returns an empty result or an error.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`QuantileError::NonFiniteValue`] if the value is not finite
+    /// when casting to a float.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oximeter::Quantile;
+    /// let mut q = Quantile::new(0.9).unwrap();
+    /// q.append(10).unwrap();
+    /// assert_eq!(q.len(), 1);
+    /// ```
+    pub fn append<T>(&mut self, value: T) -> Result<(), QuantileError>
+    where
+        T: HistogramSupport,
+    {
+        if !value.is_finite() {
+            return Err(QuantileError::NonFiniteValue);
+        }
+        // We've already checked that the value is finite.
+        let value_f = value.to_f64().unwrap();
+
+        // if !self.is_filled {
+        if self.len() < FILLED_MARKER_LEN as u64 {
+            self.marker_heights[self.len() as usize] = value_f;
+            self.marker_positions[4] += 1;
+            if self.len() == FILLED_MARKER_LEN as u64 {
+                float_ord::sort(&mut self.marker_heights);
+                self.adaptive_init();
+            }
+            return Ok(());
+        }
+
+        // Find the cell k containing the new value.
+        let k = match self.find_cell(value_f) {
+            Some(4) => {
+                self.marker_heights[4] = value_f;
+                3
+            }
+            Some(i) => i,
+            None => {
+                self.marker_heights[0] = value_f;
+                0
+            }
+        };
+
+        // Handle rounding issues as described in
+        // <https://aakinshin.net/posts/p2-quantile-estimator-rounding-issue>.
+        let count = self.len() as f64;
+        self.desired_marker_positions[1] = count * (self.p() / 2.) + 1.;
+        self.desired_marker_positions[2] = count * self.p() + 1.;
+        self.desired_marker_positions[3] = count * ((1. + self.p()) / 2.) + 1.;
+        self.desired_marker_positions[4] = count + 1.;
+
+        for i in k + 1..FILLED_MARKER_LEN {
+            self.marker_positions[i] += 1;
+        }
+
+        // Adjust height of markers adaptively to be more optimal for
+        // not just higher quantiles, but also lower ones.
+        //
+        // This is a deviation from the paper, taken from
+        // <https://aakinshin.net/posts/p2-quantile-estimator-adjusting-order>.
+        if self.p >= 0.5 {
+            for i in 1..4 {
+                self.adjust_heights(i)
+            }
+        } else {
+            for i in (1..4).rev() {
+                self.adjust_heights(i)
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Find the higher marker cell whose height is lower than the observation.
+    ///
+    /// Returns `None` if the value is less than the initial marker height.
+    fn find_cell(&mut self, value: f64) -> Option<usize> {
+        if value < self.marker_heights[0] {
```

Comment:
It feels like we're reimplementing `slice::partition_point()` but I'm not sure it's worth changing with such small arrays and the 1-based indexing differences.

Reply:
I'll double-check.

Reply:
Updated.